content/docs/reference/buildpack-api.md

```diff
@@ -194,7 +194,7 @@ The schema is as follows:
     Arbitrary data for buildpack.
 
 ## Build Plan
-The [Build Plan](https://github.com/buildpacks/spec/blob/master/buildpack.md#buildpack-plan-toml) is a document the buildpacks can use to pass information between the [detect](#bindetect) and [build](#bindetect) phases. The build plan is passed (by the lifecycle) as a parameter to the `detect` and `build` binaries of the buildpack.
+The [Build Plan](https://github.com/buildpacks/spec/blob/master/buildpack.md#build-plan-toml) is a document the buildpacks can use to pass information between the [detect](#bindetect) and [build](#bindetect) phases. The build plan is passed (by the lifecycle) as a parameter to the `detect` and `build` binaries of the buildpack.
 * During the `detect` phase, the buildpack(s) may write something it `requires` or `provides` (or both) into the Build Plan.
```

Comment:
After reading on (assuming I'm understanding correctly) it seems that `requires` is conceptually something that the app OR the buildpack can do, is this right? It might be worth calling out, that was the "ah ha" moment for me!

Reply:
Hmm. The buildpack is the only thing that writes to the Build Plan, so by that metric, it is the only thing that `requires`. However, what I tried to bring out through the examples is that the buildpack looks at the application to figure out what exactly it needs. On the simplest level, that means it looks to see if it is a `node`, `Java`, or `Go` application, and ensures that the relevant tool is there to run that application. 

Applications (or more likely, operators) will sometimes use configuration files to ensure that the buildpack will know exactly what version of something to use, or to explicitly request something that the buildpack may not be able to detect. That is kind of how the paketo-buildpacks are configured for node - node files are just plain javascript files, so it is only the presence of a `package.json` or a `yarn.lock` file which says that this is a node application. I think that the paketo-buildpacks's node buildpack doesn't look for `package.json` or `yarn.lock` files, however, because those are external to node (they come from the popular package managers, but aren't really part of the node language specification), and therefore, it would just `provide` by default, without necessary being able to know if it is `required`. 

This example would probably be simpler with a language like `go`, though, where all we'd have to do is look for the presence of a `.go` file, and if we find, the `go cnb` would install `go`. That wouldn't show just what you can do to customize the build plan as much, though... 

Let me know if that made sense!!

Reply:
Ah... this makes sense. I think I was assuming that when a buildpack requires something, it means that that something is required for the buildpack to run, vs. required for the app to run. I think in the end requires is (almost?) always about what the app needs to run. This is the wrong assumption I was making... I think you've done a good job of illustrating this :)